extjson.go

```diff
@@ -0,0 +1,1694 @@
+	err = d.readObjectTerminator()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// convertBinarySubType starts after the opening quote of the string holding hex
+// bytes of the value.
+func (d *Decoder) convertBinarySubType(out []byte, subTypeBytePos int) ([]byte, error) {
+	subTypeBytes, err := d.peekBoundedQuote(2, 3)
+	if err != nil {
+		return nil, err
+	}
+	// Go requires even digits to decode hex.
+	normalizedSubType := subTypeBytes
+	if len(normalizedSubType) == 1 {
+		normalizedSubType = []byte{'0', normalizedSubType[0]}
+	}
+	var x [1]byte
+	xs := x[0:1]
+	_, err = hex.Decode(xs, normalizedSubType)
+	if err != nil {
+		return nil, d.parseError(subTypeBytes[0], fmt.Sprintf("error parsing subtype: %v", err))
+	}
+	overwriteTypeByte(out, subTypeBytePos, xs[0])
+	_, _ = d.json.Discard(len(subTypeBytes) + 1)
+
+	return out, nil
+}
+
+// convertScope starts after the `"$scope"` key.  Having a leading $scope means
+// this is code w/ scope, but we have to buffer the scope document and write it
+// to the destination after we convert the $code string that follows.
+func (d *Decoder) convertScope(out []byte) ([]byte, error) {
+	// consume ':'
+	err := d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+
+	// Reserve length bytes for full code w/ scope length
+	cwsLengthPos := len(out)
+	out = append(out, emptyLength...)
+
+	// Copy $scope into a temporary BSON document
+	scopeDoc := make([]byte, 0, 256)
+	err = d.readCharAfterWS('{')
+	if err != nil {
+		return nil, err
+	}
+	scopeDoc, err = d.convertObject(scopeDoc, topContainer)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find and copy $code to the output
+	err = d.readCharAfterWS(',')
+	if err != nil {
+		return nil, err
+	}
+	err = d.readQuoteStart()
+	if err != nil {
+		return nil, err
+	}
+	err = d.readSpecificKey(jsonCode)
+	if err != nil {
+		return nil, err
+	}
+	err = d.readQuoteStart()
+	if err != nil {
+		return nil, err
+	}
+	out, err = d.convertString(out)
+	if err != nil {
+		return nil, err
+	}
+
+	// Write buffered $scope
+	out = append(out, scopeDoc...)
+
+	// BSON code w/scope length is total length including length bytes
+	cwsLength := len(out) - cwsLengthPos
+	overwriteLength(out, cwsLengthPos, cwsLength)
+
+	return out, nil
+}
+
+// convertRegex starts after the opening quote of `"$regex"`.  We need to
+// distinguish between Extended JSON $regex or MongoDB $regex query operator.
+// If we can peek far enough, we can check with regular expresssions.
+//
+// Both query and extended JSON allow { "$regex": "...", "$options": "..." } so
+// we choose to treat that like extended JSON.  If converted to a BSON regular
+// expression and sent as a query to a MongoDB and it will work either way.
+// However, a Javascript query expression like { "$regex": /abc/ } will turn
+// into extended JSON like { "$regex" : { <$regex or $regularExpression object>
+// } }, so if we see that "$regex" is followed by an object, we treat that as a
+// query.
```

Comment:
[nit] move this comment to the line before the element it describes (`convertRegex`)